plugins/vppv2/stnplugin/descriptor/stn.go

```diff
@@ -0,0 +1,194 @@
+// limitations under the License.
+
+//go:generate descriptor-adapter --descriptor-name STN  --value-type *stn.Rule --import "../model/stn" --output-dir "descriptor"
+
+package descriptor
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/go-errors/errors"
+	"github.com/gogo/protobuf/proto"
+	"github.com/ligato/cn-infra/logging"
+	scheduler "github.com/ligato/vpp-agent/plugins/kvscheduler/api"
+	ifDescriptor "github.com/ligato/vpp-agent/plugins/vppv2/ifplugin/descriptor"
+	"github.com/ligato/vpp-agent/plugins/vppv2/model/interfaces"
+	"github.com/ligato/vpp-agent/plugins/vppv2/model/stn"
+	"github.com/ligato/vpp-agent/plugins/vppv2/stnplugin/descriptor/adapter"
+	"github.com/ligato/vpp-agent/plugins/vppv2/stnplugin/vppcalls"
+)
+
+const (
+	// STNDescriptorName is the name of the descriptor for VPP STN rules
+	STNDescriptorName = "vpp-stn-rules"
+
+	// dependency labels
+	stnInterfaceDep = "stn-interface"
+)
+
+// A list of non-retriable errors:
+var (
+	// ErrSTNWithoutInterface is returned when VPP STN rule has undefined interface.
+	ErrSTNWithoutInterface = errors.New("VPP STN rule defined without interface")
+
+	// ErrSTNWithoutIPAddress is returned when VPP STN rule has undefined IP address.
+	ErrSTNWithoutIPAddress = errors.New("VPP STN rule defined without IP address")
+)
+
+// STNDescriptor teaches KVScheduler how to configure VPP STN rules.
+type STNDescriptor struct {
+	// dependencies
+	log        logging.Logger
+	stnHandler vppcalls.StnVppAPI
+}
+
+// NewSTNDescriptor creates a new instance of the STN descriptor.
+func NewSTNDescriptor(stnHandler vppcalls.StnVppAPI, log logging.PluginLogger) *STNDescriptor {
+	return &STNDescriptor{
+		log:        log.NewLogger("stn-descriptor"),
+		stnHandler: stnHandler,
+	}
+}
+
+// GetDescriptor returns descriptor suitable for registration (via adapter) with
+// the KVScheduler.
+func (d *STNDescriptor) GetDescriptor() *adapter.STNDescriptor {
+	return &adapter.STNDescriptor{
+		Name:               STNDescriptorName,
+		KeySelector:        d.IsSTNKey,
+		ValueTypeName:      proto.MessageName(&stn.Rule{}),
+		ValueComparator:    d.EquivalentSTNs,
+		NBKeyPrefix:        stn.Prefix,
+		Add:                d.Add,
+		Delete:             d.Delete,
+		ModifyWithRecreate: d.ModifyWithRecreate,
+		IsRetriableFailure: d.IsRetriableFailure,
+		Dependencies:       d.Dependencies,
+		Dump:               d.Dump,
+		DumpDependencies:   []string{ifDescriptor.InterfaceDescriptorName},
+	}
+}
+
+// IsSTNKey returns true if the key is identifying VPP STN rule configuration.
+func (d *STNDescriptor) IsSTNKey(key string) bool {
+	_, _, isSTNKey := stn.ParseKey(key)
+	return isSTNKey
+}
+
+// EquivalentSTNs is case-insensitive comparison function for stn.Rule.
+func (d *STNDescriptor) EquivalentSTNs(key string, oldSTN, newSTN *stn.Rule) bool {
+	// parameters compared as usually
+	if oldSTN.Interface != newSTN.Interface || oldSTN.IpAddress != newSTN.IpAddress {
```

Comment:
It's always better to use `proto.Equal` when possible and when no special behavior is required, because it's safer (you won't forget to put new fields here).